### DISCUSSION TITLE
Issue with selecting text on macOS

### DISCUSSION THREAD

**Discussion by Integralist** [ghostty-org/ghostty#3142] [Q&A] [answered]
👋

I'm on macOS and I don't know how to configure `Shift + <arrow key>` for selecting words in the command prompt's input (currently the prompt outputs `;2D` and `;2C` depending on the direction).

I tried setting a custom keybind but then discovered what I thought would be the correct combination is actually already set 🤔 so I wasn't sure if it was a bug...

```shell
$ ghostty +list-keybinds

...
shift + left                            adjust_selection:left
...
```

Also, side question, is there a simple way to select the entire prompt input line? I try `Shift+Cmd+<left-arrow>` but it prints `;10D`.

Thanks for any guidance you can give me.

**Comment by Neutrollized** [comment]
[This action does nothing if there's no active selection](https://ghostty.org/docs/config/keybind/reference#adjust_selection)

so if you highlight a bunch of text (let's say the output of `ghostty +list-keybind`, doing a shift + up for me will expand the highlighted lines upward, and your shift + left should move your cursor left accordingly.  Is this not the behavior you're experiencing?  If you having nothing highlighted, then yeah, it'll show `;2D` or whatever in the prompt

**Reply by Integralist** [reply]
OK, so yes if I select some text from the output and do `shift + up` then it adjusts the text upwards.

If I type a command into the prompt (e.g. `brew list`) and manually select some of it with the mouse (e.g. I'll double click on the word `brew` and then press `shift + right` a few times, then _eventually_ I'll have selected both words).

I guess it would be a much better user experience if a user didn't have to move their hand away from the keyboard to do text selection, and also if there were mappings for better selecting entire words or lines.

I re-tested with Terminal.app and it's the same behaviour so it's not like ghostty is doing something _wrong_ (or different) per se.

I've been using [Warp](https://www.warp.dev/) for the last couple of years and I think I've been spoilt by its ability to do easy text selection using just the keyboard. I realise now how nice that experience was 😬 

e.g. in Warp I can do `cmd + left` to move the cursor to the start of the line, then `cmd + shift + right` to select the entire line (or `alt + shift + right` to select individual words).

Is it possible to replicate that behaviour using `keybind` in ghostty?

**Reply by sakethrajesh** [reply]
I am looking for this same functionality!

**Reply by codythatsme** [reply]
> OK, so yes if I select some text from the output and do `shift + up` then it adjusts the text upwards.

Coming from Warp also and this is exactly the behaviour I wish to replicate. Essentially just the standard MacOS keyboard shortcuts for highlighting text.

**Reply by kaanmertkoc** [reply]
I tried to do it with config using claude but have no luck with selecting the texts only did word jump and whole line delete with this

```
keybind = alt+left=text:\x1b\x62
keybind = alt+right=text:\x1b\x66
keybind = cmd+backspace=text:\x15
```
I also have been using **Warp** and i really need this to switch over...

**Reply by SahSantoshh** [reply]
I am also a **Warp** user. It would be very helpful once we have this feature 🙏🏼

**Reply by MattOrmianek** [reply]
As a Warp user, I want to edit commands like text using shortcuts such as Option + Shift + Arrow keys, just as I do in Warp. This feature is a standout for me. It would be nice one we have this feature.

**Reply by MattOrmianek** [reply]
This is demo of that feature: https://docs.warp.dev/features/editor

**Reply by nashv** [reply]
> [This action does nothing if there's no active selection](https://ghostty.org/docs/config/keybind/reference#adjust_selection)
> 
> so if you highlight a bunch of text (let's say the output of `ghostty +list-keybind`, doing a shift + up for me will expand the highlighted lines upward, and your shift + left should move your cursor left accordingly. Is this not the behavior you're experiencing? If you having nothing highlighted, then yeah, it'll show `;2D` or whatever in the prompt

This is the expected behavior but it should do the same even when there is no active selection.

**Reply by duttaoindril** [reply]
Want to echo [this from hacker news](https://news.ycombinator.com/item?id=42517447#42521771):

![image](https://github.com/user-attachments/assets/eef8e61a-b545-41f2-9aaa-cacb70f3c1f4)

I’m personally used to normal macOS text editing like in Warp and any text editor, where I can select the entire command using cmd+a and easily select text with shift(+cmd)(+option) left/right.

I’d love to get this turned into an actively tracked issue.

**Reply by LinuxUserGD** [reply]
Same behaviour on Linux, see https://github.com/ghostty-org/ghostty/discussions/5499

**Comment by makzsslm** [comment]
FYI, Mitchell addressed this in the [Changelog Podcast](https://changelog.fm/622#t=1:03:16).

tl,dr: 
1. Mitch is aware of the issue 
2. It's on the roadmap
3. It's hard to fix, complex interaction between Terminal and Shell

**Reply by kaanmertkoc** [reply]
thank you for the update. didn't know it was a complex feature just took it for granted.

**Reply by 00-kat** [reply]
@kaanmertkoc it's not *necessarily* a complex feature.  Let me try to explain:
- If you want to extend a selection with keybindings, that's trivial and you already can do it.
- If you want to *start* a selection with keybindings, you need to know *where* to start it from.  The solution most want is probably to extend the selection from the current cursor position, but that can be misleading too because what many want is to select text *in the prompt* rather than *in the scrollback*.
- If you want to select the *text in your prompt* rather than *text in your scrollback*, you need the shell to bind the keypresses themselves and act on it appropriately (which is unlikely to happen in shells like Zsh and Bash outside of hopes and dreams), or you need:
  - A shell which supports selection (Bash doesn't as far as I can tell), or a way to hack on selection to a shell which doesn't support it.
  - A way to get Ghostty to tell the shell how to do it (this can be added on to the shell integration Ghostty already has).
  - Keybindings which do this from the shell side (this can be added on to the shell integration Ghostty already has).
  - People familiar with all Bash, Zsh, Fish and Elvish to add support for this to all of those shells (those four are the ones which Ghostty currently has shell integration for).
  - Ensure these new keybindings don't break Emacs (default) or Vi bindings that many other users rely on.
  - Possibly more things which I hadn't realised.

**Comment by Integralist** [comment]
Just as an update I stumbled across https://stackoverflow.com/questions/5407916/zsh-zle-shift-selection

There's an entry from 2024 which basically helped me get to a semi-working solution until official support comes.

Specifically, here is the chunk of zsh shell code I added...
https://github.com/Integralist/dotfiles/blob/628ca840549d447552311b90b3644a4a3a8208b9/.config/zsh/bindings.zsh#L59C3-L390

This allows me to...

- `Shift+LeftArrow` and `Shift+RightArrow` to select by character.
- `Shift+Ctrl+LeftArrow` and `Shift+Ctrl+RightArrow` to select by word.
- `Shift+Fn+LeftArrow` and `Shift+Fn+RightArrow` to select to start/end of line.
- `Cmd+c` to copy the selection
- `Cmd+v` to paste the selection

**Reply by Integralist** [reply]
I've spent some time reducing the code so it only does what I needed it to, and to get it closer to how macOS works.

Here's what I have working so far...

```
| BINDING                  | ACTION                 | SUPPORTED | WORK-AROUND
| Cmd + c                  | Copy selection         | ✅        |
| Cmd + v                  | Paste selection        | ✅        |
| Shift + LeftArrow        | Select left character  | ✅        |
| Shift + RightArrow       | Select right character | ✅        |
| Shift + Opt + LeftArrow  | Select word to left    | ✅        |
| Shift + Opt + RightArrow | Select word to right   | ✅        |
| Cmd + Shift + LeftArrow  | Select to line start   | ✅        |
| Cmd + Shift + RightArrow | Select to line end     | ✅        |
| Cmd + LeftArrow          | Move to line start     | ✅        |
| Cmd + RightArrow         | Move to line end       | ✅        |
| Opt + LeftArrow          | Move word to left      | ❌        | (Fn + LeftArrow)
| Opt + RightArrow         | Move word to right     | ❌        | (Fn + RightArrow)
| Opt + Backspace          | Delete word to left    | ❌        | ❌
| Opt + Fn + Backspace     | Delete word to right   | ❌        | ❌
```

> [!IMPORTANT]
> This is only for macOS and Zsh.

https://github.com/Integralist/dotfiles/blob/main/.config/zsh/bindings.zsh#L51-L208

**Reply by makzsslm** [reply]
Thanks for investigating! I'm having a couple of issues:

- `cmd + Shift + Arrow` doesn't work, writes 6~ to the shell
- I use `Right Arrow` to accept completions from [zsh-autocompletions](https://github.com/zsh-users/zsh-autosuggestions) ([Here](https://github.com/zsh-users/zsh-autosuggestions/blob/0e810e5afa27acbd074398eefbe28d13005dbc15/zsh-autosuggestions.zsh#L74C1-L74C15) is how it does the binding). Using your config, this does not work any more.
- `Opt + Arrow` and `Opt + Backspace` strangely do work to `Move word to left/right`. There might be a difference in my config.
- `Opt + Fn + Backspace` doesn't work, but `Ctrl + Fn + Backspace` does `Delete word to right`

**Reply by cbruegg** [reply]
Opt+Arrow works for me in fish shell, but not in `nano` for example :(

**Reply by Integralist** [reply]
I can't wait for this to be a solved issue by ghostty 🤞

**Reply by hugows** [reply]
Thank you this helps a lot. Another pattern I use and doesn't work here yet is "selecting some part of the input and pasting new stuff over it". :)

**Comment by 00-kat** [comment, marked as answer]
I think this discussion should be split into two.

There's a problem here: the original question is exceedingly vague, and so people who want two entirely different things end up here hoping for a solution.  I'm also not sure who's to blame, but I believe I did not help by marking both categories as a duplicate of this one.  Namely, people want:
- #6916
- #6917

And while they both sound super similar, are actually very different things.

I have already opened two such discussions, as you can see above; I believe this discussion should be closed and those two should be used to facilitate conversation on these two feature request, both of which are severely under-specified.

The only things that would be lost through this split are the previous comments here (but, I have added a note to check this discussion to both posts), and the large number of upvotes this discussion has gathered:
![A screenshot of the upvote button, highlighting 60 upvotes.](https://github.com/user-attachments/assets/c1cadb04-41fb-41eb-b7f8-d2982825e570)

**Reply by 00-kat** [reply]
Also: to anyone who sees this comment, please consider adding some suggestions for the questions I had asked in the aforementioned discussions; it would be amazing if these feature requests were to actually be brainstormed instead of just being left there indefinitely.
